### PR TITLE
Full support for CommonJS auto-imports in JS

### LINKED ISSUE #20292: Use node require for auto imports?
From https://github.com/Microsoft/vscode/issues/38773

**Repo**

Using the same project as https://github.com/Microsoft/TypeScript/issues/19629:

```js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
// @ts-check

var vscode = require('vscode');

RelativePat|
```

Accept the completion for `RelativePattern` from `vscode`

**Expected**
Respects existing module style and adds `require`:

```js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
// @ts-check

var vscode = require('vscode');
var { RelativePattern } = require('vscode');

RelativePattern
```

**Actual**
`import` inserted before everything else in the document:

```js
import { RelativePattern } from 'vscode';

// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
// @ts-check

var vscode = require('vscode');

RelativePattern
```

### LINKED ISSUE #30549: Automatic imports for react function not working in js but resolved correctly in ts
VSCode doesn't show autoimport suggestions on my machine.

I've reproduced this issue on previous stable release and on insiders version (clean install, no user settings).  `code --verbose` and console in dev. tools shows nothing suspicious. 

On launch it creates operation_log.txt in temp folder:
```
[11884:10884:0314/222311.384:VERBOSE1:crash_service_main.cc(81)] Session start. cmdline is [--reporter-url=https://rink.hockeyapp.net/api/2/apps/9dca90f9ee8449a2b45bb36238fb0879/crashes/upload --application-name=VSCode "--crashes-directory=C:\Users\Александр\AppData\Local\Temp\VSCode Crashes" --v=1]
[11884:10884:0314/222311.391:VERBOSE1:crash_service.cc(150)] window handle is 002C059C
[11884:10884:0314/222311.391:VERBOSE1:crash_service.cc(291)] pipe name is \\.\pipe\VSCode Crash Service
dumps at C:\Users\Александр\AppData\Local\Temp\VSCode Crashes
[11884:10884:0314/222311.391:VERBOSE1:crash_service.cc(295)] checkpoint is C:\Users\Александр\AppData\Local\Temp\VSCode Crashes\crash_checkpoint.txt
server is https://rink.hockeyapp.net/api/2/apps/9dca90f9ee8449a2b45bb36238fb0879/crashes/upload
maximum 128 reports/day
reporter is electron-crash-service
[11884:10884:0314/222311.393:VERBOSE1:crash_service_main.cc(95)] Ready to process crash requests
[11884:12436:0314/222311.393:VERBOSE1:crash_service.cc(325)] client start. pid = 11636
[11884:12436:0314/222311.481:VERBOSE1:crash_service.cc(325)] client start. pid = 14128
[628:9556:0314/222311.519:VERBOSE1:crash_service_main.cc(81)] Session start. cmdline is [--reporter-url=https://rink.hockeyapp.net/api/2/apps/9dca90f9ee8449a2b45bb36238fb0879/crashes/upload --application-name=VSCode "--crashes-directory=C:\Users\Александр\AppData\Local\Temp\VSCode Crashes" --v=1]
[628:9556:0314/222311.529:VERBOSE1:crash_service.cc(150)] window handle is 001504C2
[628:9556:0314/222311.530:VERBOSE1:crash_service.cc(291)] pipe name is \\.\pipe\VSCode Crash Service
dumps at C:\Users\Александр\AppData\Local\Temp\VSCode Crashes
[628:9556:0314/222311.530:VERBOSE1:crash_service.cc(295)] checkpoint is C:\Users\Александр\AppData\Local\Temp\VSCode Crashes\crash_checkpoint.txt
server is https://rink.hockeyapp.net/api/2/apps/9dca90f9ee8449a2b45bb36238fb0879/crashes/upload
maximum 128 reports/day
reporter is electron-crash-service
[628:9556:0314/222311.530:ERROR:crash_service.cc(302)] could not start dumper
[11884:13552:0314/222606.151:VERBOSE1:crash_service.cc(334)] client end. pid = 11636
[11884:13552:0314/222606.194:VERBOSE1:crash_service.cc(334)] client end. pid = 14128
[11884:13552:0314/222607.195:VERBOSE1:crash_service.cc(355)] zero clients. exiting
[11884:10884:0314/222607.195:VERBOSE1:crash_service.cc(527)] session ending..
[11884:10884:0314/222607.195:VERBOSE1:crash_service.cc(532)] clients connected :2
clients terminated :2
dumps serviced :0
dumps reported :0
[11884:10884:0314/222607.195:VERBOSE1:crash_service_main.cc(100)] Session end. return code is 0
```

And it creates a folder `vscode-typescript` with an empty file `tscancellation-6e5026c63e8646d8cd9f.tmp3`

Auto imports _does_ work as expected on a different machine. Is there anything i can try to solve this issue?

- VSCode Version: 1.32.2
- OS Version: Windows 10 Pro (1809) x64

Steps to Reproduce:

1. For example, `npm install react`. Create index.js. Start typing `useSta`. No auto import suggestions

Does this issue occur when all extensions are disabled?: Yes

### PR DESCRIPTION
**Pull request by andrewbranch**
In #32684, I added the ability for auto-imports to write a `require` variable declaration in JS files. However, it didn’t work for named or default exports, and under default project settings, would never be triggered from completions; it was only reachable by code fix. This is particularly problematic for JS, because unless the user has explicitly enabled type checking JS through VS Code settings, a tsconfig or jsconfig file, or `// @ts-check`, you can’t _get_ code fixes because you don’t get _errors_.

This PR removes those limitations, and also recognizes members of object literals assigned to `module.exports` as importable symbols.

**How do we choose between `import` and `require` for a JS file?**

_Rules are evaluated in order (highest priority first)._

1. If the imported identifier could be added to an existing `import`, add it there.
1. If the imported identifier could be added to an existing `require`, add it there.
1. If the importing file contains any ES2015+ import/export syntax, use `import`.
1. If the importing file contains any CommonJS syntax, use `require`.
1. If the compiler options indicate a module target of ES2015+, use `import`.
1. Use `require`.

**How are `require` variable declarations written for different kinds of exports?**

- `const { foo } = require('./foo')` maps to named ES exports and CommonJS exported object literal members (`module.exports = { foo }`).
- `const { default: foo } = require('./foo')` maps to default ES exports (hopefully people aren’t actually doing this).
- `const foo = require('./foo')` maps to everything else.

Fixes #20292
Fixes #30549

Related:
- #36611
- #36134

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -3273,6 +3276,10 @@ namespace ts {
                     });
                     extendExportSymbols(symbols, nestedSymbols);
                 }
+                if (symbol.escapedName === InternalSymbolName.ExportEquals && isInJSFile(symbol.valueDeclaration) && isObjectLiteralExpression((symbol.valueDeclaration as BinaryExpression).right)) {
+                    const objectLiteralSymbol = (symbol.valueDeclaration as BinaryExpression).right.symbol;
+                    extendExportSymbols(symbols, objectLiteralSymbol.members);
+                }
```
Comment (andrewbranch) [inline src/compiler/checker.ts]: I think this is probably the wrong place for this, but wanted to get some opinions. My first instinct was to make `module.exports = { foo }` recognized by the binder as something special, but that got complicated and blew up very quickly. I could just move this to `importFixes`, but was never really satisfied with having to look so closely at syntax at this point.

File: src/services/codefixes/importFixes.ts [hunk 2]
```diff
@@ -164,14 +167,14 @@ namespace ts.codefix {
...
+        readonly useRequire: boolean;
     }
 
     const enum ImportKind {
         Named,
         Default,
         Namespace,
-        Equals,
-        ConstEquals
+        CommonJS,
```
Comment (andrewbranch) [inline src/services/codefixes/importFixes.ts]: This is the fundamental conceptual change in this file, and everything else is basically just implementation details. This enum used to represent final syntax; now it’s more a representation of where the symbol can be found in the module symbol. Whether or not to use JS is then calculated and stored elsewhere, so that

- `CommonJS` can mean `import foo = require('mod')` or `const foo = require('mod')`
- `Namespace` can mean `import * as foo from 'mod'` or `const foo = require('mod')`
- `Default` can mean `import foo from 'mod'` or `const { default: foo } = require('mod')`
- `Named` can mean `import { foo } from 'mod'` or `const { foo } = require('mod')`

File: src/compiler/checker.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/checker.ts]: I'm like 99% sure that if we did something like this, it'd belong in `getCommonJsExportEquals` with our other adhoc cjsy export-assignment code, rather than in `getExportsOfModuleWorker` directly (so the result is cached and considered _before_ any potential reexports).

_However_. Today, `module.exports = { foo }` is no different than `export = { foo }` (and I don't think we should privilege one more than the other behaviorally), and `export = { 
foo }` creates an `export=` member on the containing file which is essentially just a variable declaration. While these members can be looked up as "exports" of the file, they are essentially only "members" of the declaration. The "exports" vs "members" distinction is a very internal thing that, via aliasing, isn't always clear (and is _very unfortunately_ exposed by our public `getExportsOfModule` function). However, what _is_ clear is the uniform way to get the full, final list of exported members of a module, and it's _not_ the checker-internal `getExportsOfModule` function (hilariously enough); that exists to traverse over a (admittedly flattened) representation of the exports-as-written, _without_ then layering on an `export = <expr>` resolution.

Long story short, there's a `checker.getExportsOfModule` call in `importFixes.ts` that should probably be `checker.getExportsAndPropertiesOfModule` instead (the later gets the "finalized" exports by deferring to `getPropertiesOfType(getTypeOfSymbol(moduleSymbol))`). Once you swap that (which affects no existing tests, I just checked), I'm pretty sure you shouldn't need this bit of checker code at all.

Comment (andrewbranch) [reply src/compiler/checker.ts]: Hmm, good to know that exists, and I agree there’s no reason to treat `module.exports = { foo }` differently from `exports = { foo }`. But, I’m not sure I want to get members if it’s something like

```ts
exports = new SingletonThing();
```

or something like

```ts
exports = styled.button<ThemeOne | ThemeTwo | ThemeThree | ...>`
  fontFamily: ${props => props.theme.fontFamily}
`;
```

because a) members of those would be nonsensical to auto-import, and b) this code path also happens for completions-list import suggestions, and since `getPropertiesOfType` will happily resolve all members of union and intersection types, I’m concerned that this is ripe for a 10x slowdown for everyone who happens to have `@types/coolest-most-cleverest-types-ever` installed.

I think it probably makes sense only to dive into properties if the `export=` symbol actually resolves to an _object literal_. Where do you think would make the most sense to do that?

Comment (weswigham) [reply src/compiler/checker.ts]: I don't see why
```ts
const a = {
  foo
};
module.exports = a;
```
is any different, from the perspective of our analysis, from
```ts
module.exports = {
  foo
};
```
- the first is just an alias pointing at the second. Moreover,
```ts
module.exports = (function() {
  return { foo };
})();
```
also isn't a wholly uncommon pattern and is fairly reasonable. I also remember `lighthouse` has a bunch of export assigned call expressions that do things like annotate a class with more members, which is another valid pattern.

Sure, I'd strip any autoimports from _primitives_ (or arrays) that get export-assigned (since who would want an autoimport for a string export's `valueOf`), but that can be filtered for after the fact, IMO. Pretty much any export assigned objecty type produces useful export member names; I don't think the filtering here should be syntactic in nature.

Comment (weswigham) [reply src/compiler/checker.ts]: Actually, I think a good heuristic is probably just dropping all symbols returned by `getExportsAndPropertiesOfModule` which have a `declaration` in the `lib`... technically this's mean, eg, a `Promise` polyfill which just reexports the global type doesn't get import completions, but that's probably fine the global is available).

Comment (andrewbranch) [reply src/compiler/checker.ts]: That’s fair. We already drop members of primitives, arrays, and objects that are definitely class instances (I added the latter two conditions a few months ago, in fact.) And to be fair, I think we would/could easily resolve `module.exports = foo` to `foo`’s declaration to see if _it_ was an object literal, since that’s bound as an alias. But anyway, are you not worried about the potential performance impact of running `getPropertiesOfType` on every `export=` symbol we can find?

Comment (weswigham) [reply src/compiler/checker.ts]: Pfffffffffft, `getPropertiesOfType` isn't recursive or anything, it's just getting members (and not the types of those members). _Most of the time_, it _is_ just a `symbol.exports` or `symbol.members` lookup; it just also handles all the times when that's _not_ the case (which is when it's an expression), and imo generally they're cases worth handling. It's not worse than calling `checkExpressionCached` on every `export = <expr>` position (since that's all it desugars to in the 'worst' case, essentially).

Comment (andrewbranch) [reply src/compiler/checker.ts]: FWIW I'd also be scared to call `checkExpressionCached` on every `export=`, but I’ll give it a shot 😄

File: src/services/codefixes/importFixes.ts [hunk 3]
```diff
@@ -363,21 +402,21 @@ namespace ts.codefix {
...
-    function newImportInfoFromExistingSpecifier({ declaration, importKind }: FixAddToExistingImportInfo, preferTypeOnlyImport: boolean): FixAddNewImport | undefined {
-        const expression = declaration.kind === SyntaxKind.ImportDeclaration
-            ? declaration.moduleSpecifier
-            : declaration.moduleReference.kind === SyntaxKind.ExternalModuleReference
-                ? declaration.moduleReference.expression
-                : undefined;
-        return expression && isStringLiteral(expression)
-            ? { kind: ImportFixKind.AddNew, moduleSpecifier: expression.text, importKind, typeOnly: preferTypeOnlyImport }
+    function newImportInfoFromExistingSpecifier({ declaration, importKind }: FixAddToExistingImportInfo, preferTypeOnlyImport: boolean, useRequire: boolean): FixAddNewImport | undefined {
+        const moduleSpecifier = declaration.kind === SyntaxKind.ImportDeclaration ? declaration.moduleSpecifier :
```
Comment (sandersn) [inline src/services/codefixes/importFixes.ts]: :+1: cond-style formatting

File: src/compiler/types.ts [hunk 4]
```diff
@@ -3817,6 +3818,9 @@ namespace ts {
     /* @internal */
     export type AnyImportSyntax = ImportDeclaration | ImportEqualsDeclaration;
 
+    /* @internal */
+    export type AnyImportOrRequire = AnyImportSyntax | RequireVariableDelcaration;
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/types.ts]: WOW would you look at that

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -1853,6 +1853,20 @@ namespace ts {
         return !requireStringLiteralLikeArgument || isStringLiteralLike(arg);
     }
 
+    /**
+     * Returns true if the node is a VariableDeclaration initialized to a require call (see `isRequireCall`).
+     * This function does not test if the node is in a JavaScript file or not.
+     */
+    export function isRequireVariableDeclaration(node: Node, requireStringLiteralLikeArgument: true): node is RequireVariableDelcaration;
```
Comment (andrewbranch) [inline src/compiler/utilities.ts]:
```suggestion
    export function isRequireVariableDeclaration(node: Node, requireStringLiteralLikeArgument: true): node is RequireVariableDeclaration;
```
